exercises/binary-search-tree/binary_search_tree.exs

```diff
@@ -0,0 +1,27 @@
+defmodule BinarySearchTree do
+  @type bst_node :: %{data: any, left: bst_node | nil, right: bst_node | nil}
+
+  @doc """
+  Create a new Binary Search Tree with root's value as the given 'data'
+  """
+  @spec new(any) :: bst_node
+  def new(data) do
+    # Your implementation here
+  end
+
+  @doc """
+  Creates and inserts a node with its value as 'data' into the tree.
+  """
+  @spec insert(bst_node, any) :: bst_node
+  def insert(%{data: data, left: left, right: right}, data) do
```

Comment:
Sorry, just saw this, too - we have two variables named `data` here. Maybe we don't need such specific pattern matching in the first argument? Then we can use `def insert(tree, data) do`?

Reply:
This is probably a case where I may not be writing the most idiomatic code.

If we pattern match directly in the function definition we get auto destructuring to use inside the method (we need to destructure it so we can return a map that is the existing node's data with new children). The 2nd argument of data does make the signature read a bit confusing, something like new_node_value might be better)

If we refactor to your suggestion the example impl looks something like
  def insert(tree, node_value) do
    %{data: data, left: left, right: right} = tree # not neccessary with old signature

    if node_value <= data do
      %{data: data, left: insert(left, node_value), right: right}
    else
      %{data: data, left: left, right: insert(right, node_value)}
    end
  end

_or_ instead of adding a line to destructure the three, the values could be gotten directly as tree.data, tree.left, tree.right.

I'm leaning towards your refactor since it makes the function def look cleaner and less prescriptive of the implementation. Just thought i'd let you know what was going on in my mind when I wrote it. Let me know what you think and I'll make the change.

Reply:
My thinking behind not pattern matching in the function definition here was to let the person solving this problem arrive at that step on their own. I think it _is_ a really good thing to do, but since this is the template folks will be building off of, then I like to keep it as minimal as possible.

Just so you know, the implementation (`example.exs`) and the template (in this case `binary_tree.exs`) don't need to be totally the same. They should have the same public functions that satisfy the tests, but if you want to go nuts doing other things in your implementation, you're more than welcome to! So, if you want to do that pattern matching on the first argument of `insert/2` in your implementation, you're more than welcome!

Reply:
Gotcha that makes sense. Updated to new signature then and went ahead and squashed the commit.

Reply:
Gotcha makes sense. Made the update.